interfaces/repo.go

```diff
@@ -516,3 +516,126 @@ func (r *Repository) securitySnippetsForSnap(snapName string, securitySystem Sec
+// Unknown interfaces and plugs/slots that don't validate are not added.
+// Information about those failures are returned to the caller.
+func (r *Repository) AddSnap(snapInfo *snap.Info) error {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	snapName := snapInfo.Name
+
+	for plugName := range snapInfo.Plugs {
+		if _, ok := r.plugs[snapName][plugName]; ok {
+			return fmt.Errorf("cannot add plug %s.%s: existing plug in the way", snapName, plugName)
+		}
+	}
+	for slotName := range snapInfo.Slots {
+		if _, ok := r.slots[snapName][slotName]; ok {
+			return fmt.Errorf("cannot add slot %s.%s: existing slot in the way", snapName, slotName)
+		}
+	}
+
+	var errors []error
+
+	for plugName, plugInfo := range snapInfo.Plugs {
+		iface, ok := r.ifaces[plugInfo.Interface]
+		if !ok {
+			err := fmt.Errorf("ignoring plug %s.%s, interface %s is not supported", snapName, plugName, plugInfo.Interface)
+			errors = append(errors, err)
+			continue
+		}
+		plug := &Plug{PlugInfo: plugInfo}
+		if err := iface.SanitizePlug(plug); err != nil {
+			err := fmt.Errorf("ignoring plug %s.%s, %s", snapName, plugName, err)
+			errors = append(errors, err)
+			continue
+		}
+		if r.plugs[snapName] == nil {
+			r.plugs[snapName] = make(map[string]*Plug)
+		}
+		r.plugs[snapName][plugName] = plug
+	}
+
+	for slotName, slotInfo := range snapInfo.Slots {
+		iface, ok := r.ifaces[slotInfo.Interface]
+		if !ok {
+			err := fmt.Errorf("ignoring slot %s.%s, interface %s is not supported", snapName, slotName, slotInfo.Interface)
+			errors = append(errors, err)
+			continue
+		}
+		slot := &Slot{SlotInfo: slotInfo}
+		if err := iface.SanitizeSlot(slot); err != nil {
+			err := fmt.Errorf("ignoring slot %s.%s, %s", snapName, slotName, err)
+			errors = append(errors, err)
+			continue
+		}
+		if r.slots[snapName] == nil {
+			r.slots[snapName] = make(map[string]*Slot)
+		}
+		r.slots[snapName][slotName] = slot
+	}
+
+	if len(errors) > 0 {
+		return &addSnapError{errs: errors}
+	}
+	return nil
+}
+
+// RemoveSnap removes all the plugs and slots associated with a given snap.
+//
+// This function can be used to implement snap removal or, when used along with
+// AddSnap, snap upgrade.
+//
+// RemoveSnap does not remove connections. The caller is responsible for
+// ensuring that connections are broken before calling this method. If this
+// constraint is violated then no changes are made and an error is returned.
+func (r *Repository) RemoveSnap(snapInfo *snap.Info) error {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	snapName := snapInfo.Name
+
+	for plugName, plug := range r.plugs[snapName] {
+		if len(plug.Connections) > 0 {
+			return fmt.Errorf("cannot remove connected plug %s.%s", snapName, plugName)
+		}
+	}
+	for slotName, slot := range r.slots[snapName] {
+		if len(slot.Connections) > 0 {
+			return fmt.Errorf("cannot remove connected slot %s.%s", snapName, slotName)
+		}
+	}
+
+	for _, plug := range r.plugs[snapName] {
+		// delete(r.plugs[snapInfo.Name], plugName)
```

Comment:
Because we do this operation "in bulk" just after the loop. I should probably remove this line entirely.